File: OneSWave/HalfSpace/getSWaveAmp.m
function [ampOut] = getSWaveAmp(x,y,z,vS,f,thetaW,phiW,wC)
% for a given y,z and x grids obtained using getXGrid we use this to
% get the Plane P-wave amplitude
% x, y, z are column vectors
% vS is the W-wave velocity in m/s, f is frequency in Hz
% thetaW, phiW are the directions of propagation of the wave
% wC = 'SV' or 'SH'
lambda = vS/f;
k = 2*pi/lambda;
kVec = [k*sind(thetaW)*cosd(phiW);k*sind(thetaW)*sind(phiW);k*cosd(thetaW)];
rVec = [x,y,z];
i = sqrt(-1);
kr = rVec*kVec;
ampOut = exp(-i*(kr));
if(wC=='SH')
    kUnitVec = [-sind(phiW);cosd(phiW);0];
end
if(wC=='SV')
    kUnitVec = [cosd(thetaW)*cosd(phiW);cosd(thetaW)*sind(phiW);-sind(thetaW)];
end
%kUnitVec = kVec./(sqrt(kVec(1,1)^2+kVec(2,1)^2+kVec(3,1)^2));
ampOut = [ampOut*kUnitVec(1,1),ampOut*kUnitVec(2,1),ampOut*kUnitVec(3,1)];

end

File: OneSWave/FullSpace/getSWaveAmp.py
def getSWaveAmp(x, y, z, vS, f, thetaW, phiW, wC):
   
    x, y, z = np.asarray(x), np.asarray(y), np.asarray(z)
    rVec = np.column_stack((x, y, z))

    lambda_ = vS / f
    k = 2 * np.pi / lambda_
    kx = k * np.sin(np.deg2rad(thetaW)) * np.cos(np.deg2rad(phiW))
    ky = k * np.sin(np.deg2rad(thetaW)) * np.sin(np.deg2rad(phiW))
    kz = k * np.cos(np.deg2rad(thetaW))
    kVec = np.array([kx, ky, kz])

    kr = rVec @ kVec
    amp_phase = np.exp(-1j * kr)

    if wC.upper() == 'SH':
        kUnitVec = np.array([-np.sin(np.deg2rad(phiW)),
                              np.cos(np.deg2rad(phiW)),
                              0.0])
    elif wC.upper() == 'SV':
        kUnitVec = np.array([ np.cos(np.deg2rad(thetaW)) * np.cos(np.deg2rad(phiW)),
                              np.cos(np.deg2rad(thetaW)) * np.sin(np.deg2rad(phiW)),
                             -np.sin(np.deg2rad(thetaW))])
    else:
        raise ValueError("wC must be 'SH' or 'SV'")

    ampOut = amp_phase[:, None] * kUnitVec[None, :]

    return ampOut
